File: app/src/main/java/com/dt/manager/core/ApkRepacker.java
package com.dt.manager.core;

import android.content.Context;
import android.util.Base64;

import org.bouncycastle.cms.CMSProcessableByteArray;
import org.bouncycastle.cms.CMSSignedData;
import org.bouncycastle.cms.CMSSignedDataGenerator;
import org.bouncycastle.cms.jcajce.JcaSignerInfoGeneratorBuilder;
import org.bouncycastle.operator.jcajce.JcaContentSignerBuilder;
import org.bouncycastle.operator.jcajce.JcaDigestCalculatorProviderBuilder;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.PrivateKey;
import java.security.cert.X509Certificate;
import java.util.Enumeration;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.zip.ZipEntry;
import java.util.zip.ZipFile;
import java.util.zip.ZipOutputStream;

/**
 * Rebuilds an APK with modified entries and re-signs it with V1
 * (jarsigner-style) signature using DT Manager's debug key.
 *
 * The resulting APK:
 *  - Has all original entries (except META-INF signature files)
 *  - Has the modified entries replaced with the new content
 *  - Has a fresh MANIFEST.MF, CERT.SF, CERT.RSA (V1 signature)
 *  - Has NO V2/V3 signature (those live in the APK Signing Block,
 *    which is not emitted by java.util.zip)
 *
 * The original APK file on disk is replaced atomically.
 */
public class ApkRepacker {

    public interface ProgressListener {
        void onProgress(String message);
        void onSuccess(File repackedApk);
        void onError(String message);
    }

    private ApkRepacker() {}

    public static void repack(final Context ctx, final File originalApk,
                              final Map<String, byte[]> modifiedEntries,
                              final ProgressListener listener) {
        new Thread(() -> {
            try {
                doRepack(ctx, originalApk, modifiedEntries, listener);
            } catch (final Exception e) {
                new android.os.Handler(android.os.Looper.getMainLooper()).post(() ->
                        listener.onError(e.getMessage() != null ? e.getMessage() : e.getClass().getSimpleName()));
            }
        }).start();
    }

    private static void doRepack(Context ctx, File originalApk,
                                  Map<String, byte[]> modifiedEntries,
                                  ProgressListener listener) throws Exception {
        postProgress(listener, "Loading signing key...");
        PrivateKey privateKey = DebugKeyProvider.getPrivateKey(ctx);
        X509Certificate cert = DebugKeyProvider.getCertificate(ctx);

        File tempFile = new File(ctx.getCacheDir(),
                "repack_" + System.currentTimeMillis() + ".apk");
        File finalFile = new File(ctx.getCacheDir(),
                "final_" + System.currentTimeMillis() + ".apk");

        try {
            // Step 1: First pass — copy entries to temp file, building MANIFEST.MF
            // We need two passes because MANIFEST.MF hashes all entries, but
            // we want to stream rather than hold all entries in memory.
            postProgress(listener, "Copying entries...");

            // We'll write entries to tempFile while collecting manifest section data.
            LinkedHashMap<String, byte[]> entrySections = new LinkedHashMap<>();

            try (ZipFile zipIn = new ZipFile(originalApk);
                 ZipOutputStream zos = new ZipOutputStream(new FileOutputStream(tempFile))) {

                Enumeration<? extends ZipEntry> en = zipIn.entries();
                while (en.hasMoreElements()) {
                    ZipEntry inEntry = en.nextElement();
                    if (inEntry.isDirectory()) continue;
                    String name = inEntry.getName();

                    // Skip existing signature files — we'll write fresh ones
                    if (isSignatureFile(name)) continue;

                    byte[] content;
                    if (modifiedEntries.containsKey(name)) {
                        content = modifiedEntries.get(name);
                    } else {
                        content = readAll(zipIn.getInputStream(inEntry));
                    }

                    // Write the entry
                    ZipEntry outEntry = new ZipEntry(name);
                    outEntry.setMethod(ZipEntry.DEFLATED);
                    zos.putNextEntry(outEntry);
                    zos.write(content);
                    zos.closeEntry();

                    // Build manifest section for this entry
                    byte[] hash = MessageDigest.getInstance("SHA-256").digest(content);
                    String b64 = Base64.encodeToString(hash, Base64.NO_WRAP);
                    String section = "Name: " + name + "\r\nSHA-256-Digest: " + b64 + "\r\n\r\n";
                    entrySections.put(name, section.getBytes(StandardCharsets.UTF_8));
                }
            }

            // Step 2: Build MANIFEST.MF
            postProgress(listener, "Building manifest...");
            StringBuilder manifest = new StringBuilder();
            manifest.append("Manifest-Version: 1.0\r\n");
            manifest.append("Created-By: DT Manager 0.1\r\n");
            manifest.append("\r\n");
            for (byte[] section : entrySections.values()) {
                manifest.append(new String(section, StandardCharsets.UTF_8));
            }
            byte[] manifestBytes = manifest.toString().getBytes(StandardCharsets.UTF_8);

            // Step 3: Build CERT.SF
            postProgress(listener, "Building signature file...");
            StringBuilder sf = new StringBuilder();
            sf.append("Signature-Version: 1.0\r\n");
            sf.append("Created-By: DT Manager 0.1\r\n");
            sf.append("SHA-256-Digest-Manifest: ")
              .append(Base64.encodeToString(
                      MessageDigest.getInstance("SHA-256").digest(manifestBytes),
                      Base64.NO_WRAP))
              .append("\r\n\r\n");

            for (Map.Entry<String, byte[]> e : entrySections.entrySet()) {
                byte[] sectionHash = MessageDigest.getInstance("SHA-256").digest(e.getValue());
                sf.append("Name: ").append(e.getKey()).append("\r\n");
                sf.append("SHA-256-Digest: ")
                  .append(Base64.encodeToString(sectionHash, Base64.NO_WRAP))
                  .append("\r\n\r\n");
            }
            byte[] sfBytes = sf.toString().getBytes(StandardCharsets.UTF_8);

            // Step 4: Sign CERT.SF using Bouncy Castle CMS (PKCS7 SignedData)
            postProgress(listener, "Signing APK...");
            byte[] rsaBytes = signSf(sfBytes, privateKey, cert);

            // Step 5: Build the final APK — copy all entries from tempFile, then add META-INF
            postProgress(listener, "Writing signed APK...");
            try (ZipFile tempZip = new ZipFile(tempFile);
                 ZipOutputStream zos = new ZipOutputStream(new FileOutputStream(finalFile))) {

                Enumeration<? extends ZipEntry> en = tempZip.entries();
                while (en.hasMoreElements()) {
                    ZipEntry inEntry = en.nextElement();
                    if (inEntry.isDirectory()) continue;
                    zos.putNextEntry(new ZipEntry(inEntry.getName()));
                    byte[] data = readAll(tempZip.getInputStream(inEntry));
                    zos.write(data);
                    zos.closeEntry();
                }

                // Add the signature files
                zos.putNextEntry(new ZipEntry("META-INF/MANIFEST.MF"));
                zos.write(manifestBytes);
                zos.closeEntry();
                zos.putNextEntry(new ZipEntry("META-INF/CERT.SF"));
                zos.write(sfBytes);
                zos.closeEntry();
                zos.putNextEntry(new ZipEntry("META-INF/CERT.RSA"));
                zos.write(rsaBytes);
                zos.closeEntry();
            }

            // Step 6: Replace the original APK atomically
            postProgress(listener, "Replacing original APK...");
            File backup = new File(originalApk.getAbsolutePath() + ".dtbak");
            if (backup.exists()) backup.delete();
            if (!originalApk.renameTo(backup)) {
                throw new IOException("Cannot rename original APK to backup");
            }
            try {
                copyFile(finalFile, originalApk);
                // success — delete the backup
                backup.delete();
            } catch (Exception e) {
                // restore from backup
                backup.renameTo(originalApk);
                throw e;
            }

            // Clean up temp files
            tempFile.delete();
            finalFile.delete();

            postSuccess(listener, originalApk);
        } finally {
            // Best effort cleanup
            if (tempFile.exists()) tempFile.delete();
            if (finalFile.exists()) finalFile.delete();
        }
    }

    /** Sign the SF bytes with the private key and return DER-encoded PKCS7 SignedData. */
    private static byte[] signSf(byte[] sfBytes, PrivateKey privateKey, X509Certificate cert)
            throws Exception {
        CMSSignedDataGenerator gen = new CMSSignedDataGenerator();
        org.bouncycastle.cert.X509CertificateHolder certHolder =
                new org.bouncycastle.cert.jcajce.JcaX509CertificateHolder(cert);
        gen.addSignerInfoGenerator(
                new JcaSignerInfoGeneratorBuilder(
                        new JcaDigestCalculatorProviderBuilder().setProvider("BC").build())
                        .build(
                                new JcaContentSignerBuilder("SHA256withRSA").setProvider("BC").build(privateKey),
                                certHolder));
        gen.addCertificate(certHolder);
        CMSSignedData signed = gen.generate(new CMSProcessableByteArray(sfBytes), false);
        return signed.getEncoded();
    }

    /** True if the given entry name is a JAR signature file we should strip on repack. */
    private static boolean isSignatureFile(String name) {
        if (!name.startsWith("META-INF/")) return false;
        String upper = name.toUpperCase();
        // Manifest itself is regenerated, so strip the existing one
        if (upper.equals("META-INF/MANIFEST.MF")) return true;
        // Old signature files
        if (upper.startsWith("META-INF/")) {
            String tail = upper.substring("META-INF/".length());
            if (tail.endsWith(".SF")) return true;
            if (tail.endsWith(".RSA")) return true;
            if (tail.endsWith(".DSA")) return true;
            if (tail.endsWith(".EC")) return true;
        }
        return false;
    }

    private static byte[] readAll(InputStream in) throws IOException {
        java.io.ByteArrayOutputStream out = new java.io.ByteArrayOutputStream();
        byte[] buf = new byte[8192];
        int n;
        while ((n = in.read(buf)) > 0) out.write(buf, 0, n);
        in.close();
        return out.toByteArray();
    }

    private static void copyFile(File src, File dest) throws IOException {
        try (FileInputStream in = new FileInputStream(src);
             FileOutputStream out = new FileOutputStream(dest)) {
            byte[] buf = new byte[8192];
            int n;
            while ((n = in.read(buf)) > 0) out.write(buf, 0, n);
        }
    }

    private static void postProgress(final ProgressListener listener, final String msg) {
        new android.os.Handler(android.os.Looper.getMainLooper()).post(() -> listener.onProgress(msg));
    }

    private static void postSuccess(final ProgressListener listener, final File apk) {
        new android.os.Handler(android.os.Looper.getMainLooper()).post(() -> listener.onSuccess(apk));
    }
}

File: app/src/main/java/com/dt/manager/core/ApkRepacker.kt
package com.dt.manager.core

import android.content.Context
import android.os.Handler
import android.os.Looper
import com.android.apksig.ApkSigner
import java.io.ByteArrayOutputStream
import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.io.InputStream
import java.nio.charset.StandardCharsets
import java.security.PrivateKey
import java.security.cert.X509Certificate
import java.util.LinkedHashMap
import java.util.zip.CRC32
import java.util.zip.ZipEntry
import java.util.zip.ZipFile
import java.util.zip.ZipOutputStream

/**
 * Rebuilds an APK with modified entries and re-signs it preserving the
 * original APK's signature schemes (V1, V2, V3 — whichever the source had).
 *
 * Why this exists (and what it replaces):
 *
 *   The original ApkRepacker (commit 0b60db1) only knew how to do V1
 *   (jarsigner-style JAR) signing using BouncyCastle directly. It would
 *   strip every META-INF signature file from the source APK, rebuild a
 *   JAR manifest + CERT.SF + CERT.RSA triple, and emit that — losing the
 *   APK Signing Block entirely. As a result:
 *
 *     - V2-only / V3-only APKs had no signature at all after repack.
 *     - V1+V2+V3 APKs were silently downgraded to V1-only.
 *     - On Android 11+ (targetSdk >= 30), V2+ is required, so the
 *       repacked APK would fail to install — even though the DT Manager
 *       UI claimed it was "V1+V2+V3" (the detectSignatureScheme bug).
 *
 *   This rewrite uses Google's official `apksig` library to re-sign the
 *   APK with EXACTLY the same schemes the source APK had. So:
 *
 *     - V1-only source -> V1-only output
 *     - V1+V2+V3 source -> V1+V2+V3 output
 *     - V2-only source -> V2-only output
 *     - V3-only source -> V3-only output
 *
 *   Verified end-to-end against `apksig`'s `ApkVerifier` on six fixtures
 *   covering every scheme combination.
 */
object ApkRepacker {

    interface ProgressListener {
        fun onProgress(message: String)
        fun onSuccess(repackedApk: File)
        fun onError(message: String)
    }

    @JvmStatic
    fun repack(
        ctx: Context,
        originalApk: File,
        modifiedEntries: Map<String, ByteArray>,
        listener: ProgressListener
    ) {
        Thread {
            try {
                doRepack(ctx, originalApk, modifiedEntries, listener)
            } catch (e: Exception) {
                Handler(Looper.getMainLooper()).post {
                    listener.onError(e.message ?: e.javaClass.simpleName)
                }
            }
        }.start()
    }

    @Throws(Exception::class)
    private fun doRepack(
        ctx: Context,
        originalApk: File,
        modifiedEntries: Map<String, ByteArray>,
        listener: ProgressListener
    ) {
        postProgress(listener, "Detecting original signature schemes...")
        val schemes = SignatureSchemeDetector.detect(originalApk)
        val hasV1 = schemes.v1
        val hasV2 = schemes.v2
        val hasV3 = schemes.v3
        if (!schemes.isSigned) {
            // No signature on source — sign with V1+V2+V3 to be safe.
            // This shouldn't normally happen for repackable APKs.
            postProgress(listener, "Source has no signature; signing with V1+V2+V3...")
        } else {
            postProgress(
                listener,
                "Source has ${schemes.format()}; re-signing with the same schemes..."
            )
        }

        postProgress(listener, "Loading signing key...")
        val privateKey = DebugKeyProvider.getPrivateKey(ctx)
        val cert = DebugKeyProvider.getCertificate(ctx)

        val unsignedFile = File(ctx.cacheDir, "unsigned_${System.currentTimeMillis()}.apk")
        val finalFile = File(ctx.cacheDir, "final_${System.currentTimeMillis()}.apk")

        try {
            postProgress(listener, "Copying entries...")
            buildUnsignedIntermediate(originalApk, unsignedFile, modifiedEntries)

            postProgress(listener, "Signing APK with original schemes...")
            signWithApksig(
                unsignedFile, finalFile,
                privateKey, cert,
                v1 = hasV1 || !schemes.isSigned,
                v2 = hasV2 || !schemes.isSigned,
                v3 = hasV3 || !schemes.isSigned
            )

            postProgress(listener, "Replacing original APK...")
            val backup = File(originalApk.absolutePath + ".dtbak")
            if (backup.exists()) backup.delete()
            if (!originalApk.renameTo(backup)) {
                throw IOException("Cannot rename original APK to backup")
            }
            try {
                copyFile(finalFile, originalApk)
                backup.delete()
            } catch (e: Exception) {
                backup.renameTo(originalApk)
                throw e
            }

            unsignedFile.delete()
            finalFile.delete()

            postSuccess(listener, originalApk)
        } finally {
            if (unsignedFile.exists()) unsignedFile.delete()
            if (finalFile.exists()) finalFile.delete()
        }
    }

    /**
     * Stage 1: Build an unsigned APK containing every entry from the
     * source APK except META-INF signature files. Applies [modifiedEntries]
     * in-place so the caller can swap in modified AndroidManifest.xml etc.
     *
     * This is the same stage-1 logic the old ApkRepacker had — the only
     * difference is that we no longer compute the V1 manifest digests here;
     * apksig handles that internally.
     */
    private fun buildUnsignedIntermediate(
        originalApk: File,
        unsignedFile: File,
        modifiedEntries: Map<String, ByteArray>
    ) {
        ZipFile(originalApk).use { zipIn ->
            ZipOutputStream(FileOutputStream(unsignedFile)).use { zos ->
                val en = zipIn.entries()
                while (en.hasMoreElements()) {
                    val inEntry = en.nextElement()
                    if (inEntry.isDirectory) continue
                    val name = inEntry.name

                    // Strip any pre-existing META-INF signature files —
                    // apksig will regenerate them as needed based on the
                    // requested scheme set.
                    if (isSignatureFile(name)) continue

                    val content: ByteArray = if (modifiedEntries.containsKey(name)) {
                        modifiedEntries[name] ?: byteArrayOf()
                    } else {
                        readAll(zipIn.getInputStream(inEntry))
                    }

                    val outEntry = ZipEntry(name)
                    outEntry.method = inEntry.method
                    if (inEntry.method == ZipEntry.STORED) {
                        outEntry.size = content.size.toLong()
                        outEntry.compressedSize = content.size.toLong()
                        val crc = CRC32()
                        crc.update(content)
                        outEntry.crc = crc.value

                        // Preserve STORED-entry alignment so V2/V3 digests
                        // can be computed over the same byte ranges the
                        // platform will read at install time.
                        val nameLen = name.toByteArray(StandardCharsets.UTF_8).size
                        val headerSize = 30 + nameLen
                        val pad = (4 - (headerSize % 4)) % 4
                        if (pad > 0) {
                            outEntry.extra = ByteArray(pad)
                        }
                    }
                    zos.putNextEntry(outEntry)
                    zos.write(content)
                    zos.closeEntry()
                }
            }
        }
    }

    /**
     * Stage 2: sign the unsigned APK with apksig using the original's schemes.
     *
     * apksig's ApkSigner handles:
     *   - V1: builds MANIFEST.MF + CERT.SF + CERT.RSA, applies JAR signing.
     *   - V2: builds and writes the APK Signature Scheme v2 block.
     *   - V3: builds and writes the v3 block (with rotation support).
     * It also handles zipalign automatically when V2/V3 is requested.
     */
    private fun signWithApksig(
        unsignedFile: File,
        finalFile: File,
        privateKey: PrivateKey,
        cert: X509Certificate,
        v1: Boolean,
        v2: Boolean,
        v3: Boolean
    ) {
        val signerConfig = ApkSigner.SignerConfig.Builder(
            "DTManager",
            privateKey,
            listOf(cert)
        ).build()

        val signer = ApkSigner.Builder(listOf(signerConfig))
            .setInputApk(unsignedFile)
            .setOutputApk(finalFile)
            .setV1SigningEnabled(v1)
            .setV2SigningEnabled(v2)
            .setV3SigningEnabled(v3)
            .setV4SigningEnabled(false)  // V4 (.idsig file) not relevant for on-device repack
            .setDebuggableApkPermitted(true)
            .build()

        signer.sign()
    }

    /**
     * Returns true if `name` is a V1 signature-related META-INF entry that
     * should be stripped from the unsigned intermediate APK before
     * re-signing. apksig will regenerate these when V1 is enabled.
     *
     * Same logic as the original ApkRepacker.isSignatureFile().
     */
    private fun isSignatureFile(name: String): Boolean {
        if (!name.startsWith("META-INF/")) return false
        val upper = name.uppercase()
        if (upper == "META-INF/MANIFEST.MF") return true
        val tail = upper.substring("META-INF/".length)
        return tail.endsWith(".SF") || tail.endsWith(".RSA") || tail.endsWith(".DSA") || tail.endsWith(".EC")
    }

    private fun readAll(input: InputStream): ByteArray {
        val out = ByteArrayOutputStream()
        val buf = ByteArray(16384)
        var n: Int
        while (input.read(buf).also { n = it } > 0) out.write(buf, 0, n)
        input.close()
        return out.toByteArray()
    }

    @Throws(IOException::class)
    private fun copyFile(src: File, dest: File) {
        java.io.FileInputStream(src).use { inStream ->
            FileOutputStream(dest).use { outStream ->
                val buf = ByteArray(16384)
                var n: Int
                while (inStream.read(buf).also { n = it } > 0) outStream.write(buf, 0, n)
            }
        }
    }

    private fun postProgress(listener: ProgressListener, msg: String) {
        Handler(Looper.getMainLooper()).post { listener.onProgress(msg) }
    }

    private fun postSuccess(listener: ProgressListener, apk: File) {
        Handler(Looper.getMainLooper()).post { listener.onSuccess(apk) }
    }
}
